packages/core/helper-plugin/src/features/Tracking.js

```diff
@@ -0,0 +1,139 @@
+import * as React from 'react';
+
+import axios from 'axios';
+import PropTypes from 'prop-types';
+
+import { useAppInfo } from './AppInfo';
+
+/**
+ * @preserve
+ * @typedef {Object} TelemetryProperties
+ * @property {boolean} useTypescriptOnServer
+ * @property {boolean} useTypescriptOnAdmin
+ * @property {boolean} isHostedOnStrapiCloud
+ * @property {number} numberOfAllContentTypes
+ * @property {number} numberOfComponents
+ * @property {number} numberOfDynamicZones
+ */
+
+/**
+ * @preserve
+ * @typedef {Object} TrackingContextValue
+ * @property {string | boolean} uuid
+ * @property {string | undefined} deviceId
+ * @property {TelemetryProperties} telemetryProperties
+ */
+
+/* -------------------------------------------------------------------------------------------------
+ * Context
+ * -----------------------------------------------------------------------------------------------*/
+
+/**
+ * @preserve
+ * @type {React.Context<TrackingContextValue>}
+ */
+const TrackingContext = React.createContext();
+
+/* -------------------------------------------------------------------------------------------------
+ * Provider
+ * -----------------------------------------------------------------------------------------------*/
+
+const TrackingProvider = ({ value, children }) => {
+  const memoizedValue = React.useMemo(() => value, [value]);
+
+  return <TrackingContext.Provider value={memoizedValue}>{children}</TrackingContext.Provider>;
+};
+
+TrackingProvider.propTypes = {
+  children: PropTypes.node.isRequired,
+  value: PropTypes.shape({
+    uuid: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
+    deviceId: PropTypes.string,
+    telemetryProperties: PropTypes.object,
+  }),
+};
+
+TrackingProvider.defaultProps = {
+  value: {
+    deviceId: undefined,
+    uuid: false,
+    telemetryProperties: undefined,
+  },
+};
+
+/* -------------------------------------------------------------------------------------------------
+ * Hook
+ * -----------------------------------------------------------------------------------------------*/
+
+/**
+ * @preserve
+ * @typedef {(event: string, properties: Record<string, any>) => Promise<void>} TrackUsageFn
+ */
+
+/**
+ * @preserve
+ * @description Used to send amplitude events to the Strapi Tracking hub.
+ *
+ * @returns {{trackUsage: TrackUsageFn}}
+ *
+ * @example
+ * ```tsx
+ * import { useTracking } from '@strapi/helper-plugin';
+ *
+ * const MyComponent = () => {
+ *  const { trackUsage } = useTracking();
+ *
+ *  const handleClick = () => {
+ *   trackUsage('my-event', { myProperty: 'myValue' });
+ *  }
+ *
+ *  return <button onClick={handleClick}>Send Event</button>
+ * }
+ * ```
+ */
+const useTracking = () => {
+  const { uuid, telemetryProperties, deviceId } = React.useContext(TrackingContext);
+  const appInfo = useAppInfo();
+  const userId = appInfo?.userId;
+
+  /**
+   * @type {TrackUsageFn}
+   */
+  const trackUsage = React.useCallback(
```

Comment:
I think the fact it is wrapped in `useCallback` is new isn't it? Do you think it is save? I am not sure if we want to send one event per re-render in some situations

Reply:
The function will be reconstructed if `deviceId || telemetryProperties || userId || uuid` changes, many of which come from the context or `useAppInfo` hook – in short all of them come from very high up in the tree. So this function is unlikely to be "remade" it won't run the function just because it's re-rendered though :) so yes, I think it's safe.